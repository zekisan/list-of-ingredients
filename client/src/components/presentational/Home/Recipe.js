import React from 'react';
import PropTypes from 'prop-types';
import { Container, Header, Segment, Button } from 'semantic-ui-react';

const Recipe = ({ drink }) => <span>
  {drink &&
    <Container>
      <Header as='h2'>{drink.title}</Header>
      {drink.description && <p>{drink.description}</p>}
      {drink.ingredients &&
        <Segment.Group>
          {drink.ingredients.map((ingredient, i) => <Segment key={i}>{ingredient.description}</Segment>)}
        </Segment.Group>
      }
      {drink.steps && <p>{drink.steps}</p>}
      {drink.source && <Button basic size='tiny' color='teal' href={drink.source}>Source</Button>}
    </Container>
  }
</span>;

Recipe.propTypes = {
  /** the drink data */
  drink: PropTypes.object,
};

export default Recipe;