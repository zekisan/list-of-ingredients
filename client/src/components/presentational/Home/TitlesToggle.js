import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Container, Button } from 'semantic-ui-react';

const TitlesToggle = ({ drinks, drink, getDrink }) =>
  <span>
    {
      drinks && drinks.length
        ? <Button.Group color='teal' fluid widths={drinks.length}>
          {Object.keys(drinks).map((key) => {
            return <TitleButton
              active={drink && drink.id === drinks[key].id}
              key={key}
              drinkId={drinks[key].id}
              getDrink={getDrink}
              title={drinks[key].title}
            />
          })}
        </Button.Group>
        : <Container textAlign='center'>No drinks found.</Container>
    }
  </span>;

TitlesToggle.propTypes = {
  drinks: PropTypes.array,
  drink: PropTypes.object,
  getDrink: PropTypes.func,
};

export default TitlesToggle;

class TitleButton extends PureComponent {
  constructor(props) {
    super(props)

    this.getDrink = this.getDrink.bind(this);
  }

  getDrink() {
    this.props.getDrink(this.props.drinkId);
  }

  render() {
    const { active, title } = this.props;
    return <Button active={active} onClick={this.getDrink}>
      {title}
    </Button>
  }
}

TitleButton.propTypes = {
  active: PropTypes.bool,
  drinkId: PropTypes.number,
  title: PropTypes.string,
  getDrink: PropTypes.func,
}
