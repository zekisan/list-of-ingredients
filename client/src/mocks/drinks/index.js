const drinks = [
    {"id":1,"title":"Two-Minute Breakfast Boost","description":"Whizz up a low-fat breakfast smoothie in no time. Use banana with other soft fruit, plus honey for a little sweetness and oats for slow-release fuel.","steps":"Put all the ingredients in a blender and whizz for 1 min until smooth. Pour the mixture into two glasses to serve.","source":"https://www.bbcgoodfood.com/recipes/two-minute-breakfast-smoothie","created_at":"2018-06-09T00:22:22.524Z","updated_at":"2018-06-09T00:22:22.524Z","ingredients":[{"id":1,"description":"1 banana"},{"id":2,"description":"1 tbsp porridge oats"},{"id":3,"description":"80g soft fruit (like mango or strawberries)"},{"id":4,"description":"150ml milk"},{"id":5,"description":"1 tsp honey"},{"id":6,"description":"1 tsp vanilla extract"}]},
    {"id":2,"title":"Kale And Hearty Smoothie","description":"Give yourself a dose of vitamin C in the morning with this vegan green smoothie. Along with kale and avocado, there's a hit of zesty lime and pineapple.","steps":"Put all of the ingredients into a bullet or smoothie maker, add a large splash of water and blitz. Add more water until you have the desired consistency.","source":"https://www.bbcgoodfood.com/recipes/kale-smoothie","created_at":"2018-06-09T00:22:22.993Z","updated_at":"2018-06-09T00:22:22.993Z","ingredients":[{"id":7,"description":"2 handfuls kale"},{"id":8,"description":"½ avocado"},{"id":9,"description":"½ lime, juice only"},{"id":10,"description":"large handful frozen pineapple chunks"},{"id":11,"description":"medium-sized chunk ginger"},{"id":12,"description":"1 tbsp cashew nuts"},{"id":13,"description":"1 banana, optional"}]}
]

export default drinks;