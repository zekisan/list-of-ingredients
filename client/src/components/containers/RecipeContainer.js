import React, { PureComponent } from 'react';
import '../../App.css';
import { Dimmer, Loader, Divider } from 'semantic-ui-react';

import TitlesToggle from '../../components/presentational/Home/TitlesToggle';
import Recipe from '../../components/presentational/Home/Recipe';

class Home extends PureComponent {
  constructor() {
    super();

    this.state = {}
    this.getDrinks = this.getDrinks.bind(this);
    this.getDrink = this.getDrink.bind(this);
  }

  componentDidMount() {
    this.getDrinks();
  }

  fetch(endpoint) {
    return window.fetch(endpoint)
      .then(response => response.json())
      .catch(error => console.log(error));
  }

  getDrinks() {
    this.fetch('/api/drinks')
      .then(drinks => {
        if (drinks.length) {
          this.setState({ drinks: drinks })
          this.getDrink(drinks[0].id)
        } else {
          this.setState({ drinks: [] })
        }
      })
  }

  getDrink(id) {
    this.fetch(`/api/drinks/${id}`)
      .then(drink => this.setState({ drink: drink }))
  }

  render() {
    let { drinks, drink } = this.state;
    return <span>
      {
        drinks
          ? <span>
            <TitlesToggle drinks={drinks} drink={drink} getDrink={this.getDrink} />
            <Divider section />
            <Recipe drink={drink} />
          </span>
          :
          <Dimmer active inverted>
            <Loader content='loading' />
          </Dimmer>
      }
    </span>
  }
}

export default Home;
