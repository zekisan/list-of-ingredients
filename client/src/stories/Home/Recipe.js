import React from 'react';

import { storiesOf } from '@storybook/react';
import Recipe from '../../components/presentational/Home/Recipe';

import drinks from '../../mocks/drinks'

storiesOf('Recipe', module).add('to Storybook', () => <Recipe drink={drinks[0]}/>);