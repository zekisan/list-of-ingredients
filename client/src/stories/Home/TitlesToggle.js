import React from 'react';

import { storiesOf } from '@storybook/react';
import TitlesToggle from '../../components/presentational/Home/TitlesToggle';

import drinks from '../../mocks/drinks'

storiesOf('TitlesToggle', module).add('to Storybook', () => <TitlesToggle drinks={drinks}/>);