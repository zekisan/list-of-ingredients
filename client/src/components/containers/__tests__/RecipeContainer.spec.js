import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Dimmer, Loader, Divider } from 'semantic-ui-react';

import RecipeContainer from '../RecipeContainer';
import TitlesToggle from '../../presentational/Home/TitlesToggle';
import Recipe from '../../presentational/Home/Recipe';
import drinks from '../../../mocks/drinks'

describe('RecipeContainer', () => {
    beforeEach(() => {
        fetch.resetMocks();
    })

    describe('when state `drinks` receives data', () => {
        let wrapper;

        beforeEach(() => {
            fetch.mockResponse(JSON.stringify({ drinks: drinks }));
            wrapper = mount(<RecipeContainer />);
            wrapper.setState({ drinks: drinks })
        });

        it('always renders a span', () => {
            const spans = wrapper.find('span');
            expect(spans.length).toBeGreaterThan(0);
        });

        it('always renders a TitlesToggle', () => {
            expect(wrapper.find(TitlesToggle).length).toBe(1)
        });

        it('always renders a Divider', () => {
            expect(wrapper.find(Divider).length).toBe(1);
        });

        it('always renders a Recipe', () => {
            expect(wrapper.find(Recipe).length).toBe(1);
        });

        it('never renders a Dimmer', () => {
            expect(wrapper.find(Dimmer).length).toBe(0);
        });

        it('never renders a Loader', () => {
            expect(wrapper.find(Loader).length).toBe(0);
        });
    });

    describe('when state `drinks` receives undefined', () => {
        let wrapper;

        beforeEach(() => {
            fetch.mockResponse(JSON.stringify({ drinks: undefined }))
            wrapper = shallow(<RecipeContainer />);
        });

        it('always renders a span', () => {
            const spans = wrapper.find('span');
            expect(spans.length).toBe(1);
        });

        it('never renders a TitlesToggle', () => {
            expect(wrapper.find(TitlesToggle).length).toBe(0)
        });

        it('never renders a Divider', () => {
            expect(wrapper.find(Divider).length).toBe(0);
        });

        it('never renders a Recipe', () => {
            expect(wrapper.find(Recipe).length).toBe(0);
        });

        it('always renders a Dimmer', () => {
            expect(wrapper.find(Dimmer).length).toBe(1);
        });

        it('always renders a Loader', () => {
            expect(wrapper.find(Loader).length).toBe(1);
        });
    });

    describe('snapshot', () => {
        beforeEach(() => {
            fetch.mockResponse(JSON.stringify({ drinks: drinks }))
        });

        it('tries snapshots', () => {
            const tree = renderer
                .create(<RecipeContainer />)
                .toJSON();
            
            expect(tree).toMatchSnapshot();
        });
    });
});