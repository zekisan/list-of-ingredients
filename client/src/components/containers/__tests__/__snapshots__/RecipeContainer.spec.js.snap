// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeContainer snapshot tries snapshots 1`] = `
<span>
  <div
    className="ui active transition visible inverted dimmer"
    onClick={[Function]}
  >
    <div
      className="content"
    >
      <div
        className="ui text loader"
      >
        loading
      </div>
    </div>
  </div>
</span>
`;
